# [LAB-04] 5. 데이터 병합 - 연습문제

## 패키지 참조

In [1]:
from jussam import load_data
from pandas import DataFrame
from pandas import merge, concat

📦 연세대학교 주영아 교수가 제작한 라이브러리를 사용중입니다.
📧 Email(1): j.purplerose@yonsei.ac.kr
📧 Email(2): j.purplerose@gmail.com
📝 Website: https://juyounga.kr/


## 연습문제 1 - 신입 데이터 분석가를 위한 실전 미션!

### 고객 데이터 (직접 생성)

In [2]:
customer_data = {
    'customer_id': [101, 102, 103, 104, 105, 106],
    'name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Frank'],
    'age': [25, 32, 45, 38, 29, 41]
}
customers_df = DataFrame(customer_data)
customers_df

,customer_id,name,age
0,101,Alice,25
1,102,Bob,32
2,103,Charlie,45
3,104,David,38
4,105,Eve,29
5,106,Frank,41


### 구매 데이터 (데이터 가져오기)

In [3]:
purchases_df = load_data('order_history')
purchases_df

📚 신입 데이터 분석가를 위한 실전 미션!

    field        description
--  -----------  ----------------
 0  purchase_id  일련번호(인덱스)
 1  customer_id  고객번호
 2  item         상품명
 3  price        금액



,customer_id,item,price
purchase_id,,,
1,101,Laptop,1200.000
2,102,Mouse,25.000
3,103,Keyboard,NaN
4,101,Monitor,300.000
5,104,Webcam,5000.000
6,102,USB Hub,30.000
7,105,Laptop,1100.000
8,103,Mouse,NaN
9,101,External HDD,80.000


### 1. 구매 데이터의 price 컬럼에 있는 결측치는 전체 상품의 평균 가격으로 채워주세요.

In [4]:
# 원본 복사
결측치처리 = purchases_df.copy()

# 전체 상품의 평균 가격
average_price = 결측치처리['price'].mean()

# 평균 가격으로 결측치 대체
# --> 특정 컬럼에 대한 처리이므로 결과는 Series 형태로 반환
결측치처리_결과 = 결측치처리['price'].fillna(average_price)

# Series 객체를 원본 데이터프레임의 price 컬럼에 대입
결측치처리['price'] = 결측치처리_결과
결측치처리

,customer_id,item,price
purchase_id,,,
1,101,Laptop,1200.000
2,102,Mouse,25.000
3,103,Keyboard,985.625
4,101,Monitor,300.000
5,104,Webcam,5000.000
6,102,USB Hub,30.000
7,105,Laptop,1100.000
8,103,Mouse,985.625
9,101,External HDD,80.000


### 2. 정제된 두 데이터프레임을 customer_id 기준으로 병합하세요.

In [5]:
병합 = merge(결측치처리, customers_df, on='customer_id')
병합

,customer_id,item,price,name,age
0,101,Laptop,1200.000,Alice,25
1,102,Mouse,25.000,Bob,32
2,103,Keyboard,985.625,Charlie,45
3,101,Monitor,300.000,Alice,25
4,104,Webcam,5000.000,David,38
5,102,USB Hub,30.000,Bob,32
6,105,Laptop,1100.000,Eve,29
7,103,Mouse,985.625,Charlie,45
8,101,External HDD,80.000,Alice,25
9,106,Docking Station,150.000,Frank,41


### 3. 기술통계량 도출

- customer_id를 제외하고 기술통계량을 도출해야 한다.

In [6]:
기술통계_df = 병합.drop(columns=['customer_id', 'item', 'name'])
기술통계_df.describe().T

,count,mean,std,min,25%,50%,75%,max
price,10.000,985.625,1490.666,25.000,97.500,642.812,1071.406,5000.000
age,10.000,33.700,8.042,25.000,26.000,32.000,40.250,45.000


## 연습문제 2 - 흩어진 판매 데이터 통합하기

### 데이터 가져오기

In [7]:
강남점 = load_data('shop_강남')
강남점

📚 흩어진 판매 데이터 통합하기(인덱스/메타데이터 없음)


,OrderID,Product,Quantity,Price
0,A001,Keyboard,10,15000.000
1,A002,Mouse,20,NaN
2,A003,Monitor,5,75000.000
3,A004,Keyboard,12,15000.000


In [8]:
종로점 = load_data('shop_종로')
종로점

📚 흩어진 판매 데이터 통합하기(인덱스/메타데이터 없음)


,OrderID,Product,Quantity,Price
0,B001,Mouse,15,8000
1,B002,Keyboard,999,14500
2,B003,Webcam,10,30000
3,B004,Mouse,18,8000


### 1. 강남점 데이터에는 'Mouse' 상품의 가격(Price)이 누락되어 있습니다. 이 값을 8000으로 채워주세요.

In [9]:
강남점cp = 강남점.copy()
강남점cp.loc[1, 'Price'] = 8000
강남점cp

,OrderID,Product,Quantity,Price
0,A001,Keyboard,10,15000.000
1,A002,Mouse,20,8000.000
2,A003,Monitor,5,75000.000
3,A004,Keyboard,12,15000.000


### 2. 종로점 데이터에는 'Keyboard' 상품의 주문 수량(Quantity)이 999로 잘못 입력되어 있습니다. 실제 수량은 9이므로, 이 값을 수정해주세요.

In [10]:
종로점cp = 종로점.copy()
종로점cp.loc[1, 'Quantity'] = 9
종로점cp

,OrderID,Product,Quantity,Price
0,B001,Mouse,15,8000
1,B002,Keyboard,9,14500
2,B003,Webcam,10,30000
3,B004,Mouse,18,8000


### 3, 4. 두 개의 DataFrame을 하나의 DataFrame으로 합치세요. 합치는 과정에서 기존의 인덱스는 의미가 없으므로, 새로운 인덱스를 부여하세요. 화면에 최종 데이터 프레임을 출력하세요.

In [11]:
df_total = concat([강남점cp, 종로점cp], ignore_index=True)
df_total

,OrderID,Product,Quantity,Price
0,A001,Keyboard,10,15000.000
1,A002,Mouse,20,8000.000
2,A003,Monitor,5,75000.000
3,A004,Keyboard,12,15000.000
4,B001,Mouse,15,8000.000
5,B002,Keyboard,9,14500.000
6,B003,Webcam,10,30000.000
7,B004,Mouse,18,8000.000
